README
NAME
    Soar::Production - REPRESENT SOAR PRODUCTIONS

VERSION
    version 0.02

SYNOPSIS
DESCRIPTION
NAME
    Soar::Production- REPRESENTS A SOAR PRODUCTION

METHODS
  "new"
    Argument: text of a Soar production. Creates a new production object
    using the input text.

  "prods_from"
    This method extracts productions from a given text. It returns a
    reference to an array containing production objects. Note that all
    comments are removed as a preprocessing step to detecting and extracting
    productions. It takes a set of named arguments: 'file'- the name of a
    file to read. 'text'- the text to read. You must choose to export this
    function via the "use" function:

            use Soar::Production qw(prods_from);

  "prods_from_file"
    A shortcut for "prods_from(file =" $arg)>.

  "name"
    Optional argument: name to assign production. Sets the name of the
    current production if an argument is given. Returns the name of the
    production.

TODO
  "state_name"
    Set/get name of matched state

  "superstate_name"
    Set/get name of matched state's superstate

  "type"
    Does this production match a state or an impasse?

  "validate"
    Check this production against a datamap.

  check semantic correctness
    Soar::Production::Parser does not check semantic correctness. The
    following are good things to check:

    everything matched in RHS must be in LHS
    no empty RHS
    Only allowable non-operator preference is REJECT
    Check for existence of RHS function
    <s> not connected
    disconnect from goal or impasses (no 'state' or 'impasse' keyword)

AUTHOR
    Nathan Glenn <garfieldnate@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Nathan Glenn.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.

POD ERRORS
    Hey! The above document had some coding errors, which are explained
    below:

    Around line 147:
        Unknown directive: =over3

    Around line 149:
        '=item' outside of any '=over'



Changes

Revision history for Soar-Production
 
0.02
	Released: 2013-02-09 08:27:16 UTC
	Fixed POD formatting.

0.01
	Released: 2013-01-21 05:48:37 UTC
	Initial release; combined Soar::Production::Parser and Soar::Production::Printer.

t/parser/TestSoarProdParser.pm
#
# This file is part of Soar-Production
#
# This software is copyright (c) 2012 by Nathan Glenn.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package t::parser::TestSoarProdParser;
use Test::Base -Base;
use Soar::Production::Parser;

1;

package t::parser::TestSoarProdParser::Filter;
use Test::Base::Filter -base;
use Soar::Production::Parser;
use Data::Dumper;

my $parser = Soar::Production::Parser->new();

#return the parse of the text
sub parse { 
	# print "parse " . $_[0] . "\n"; 
	$parser->parse_text($_[0]);
}

#return 0 if the parse was unsuccessful, 1 if it was.	
sub parse_success {
	return 1 if defined $parser->parse_text($_[0]);
	return 0;
}

#an argument like LHS,stateImpCond,attrValueTests,0 will return
# $root->{LHS}->{stateImpCond}->{attrValueTests}->[0]
sub dive {
	my ($root) = shift @_;
	my @path = split ',', Test::Base::Filter::current_arguments();
	use Data::Diver qw (Dive);
	return Dive($root,@path);
}

#returns the number of productions found in a string
sub split_prods {
	my ($prods) = shift @_;
	my $productions = $parser->productions(text => $prods, parse => 0);
	# print $productions;
	return scalar @$productions;
}

#returns productions found in a string
sub get_prods {
	my ($prods) = shift @_;
	# print 'Getting prods from ' . $prods . "\n";
	my $args = Test::Base::Filter::current_arguments();
	my $productions = $parser->productions(text => $prods, parse => 0);
	# print "found $#$productions productions\n";
	return $productions
		unless defined $args;
		
	return $$productions[$args];
}

sub no_comment {
	my ($text) = @_;
	$text = $parser->no_comment($text);
	# print "returning: \n$text\n";
	return $text;
}

1;


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Data::Diver" => 0,
    "File::Find" => 0,
    "File::Spec" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Module::Build" => "0.3601",
    "Test::Base" => 0,
    "Test::Base::Filter" => 0,
    "Test::Deep" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.88",
    "Test::Warn" => 0,
    "utf8" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "REPRESENT SOAR PRODUCTIONS\r",
  "dist_author" => [
    "Nathan Glenn <garfieldnate\@gmail.com>"
  ],
  "dist_name" => "Soar-Production",
  "dist_version" => "0.02",
  "license" => "perl",
  "module_name" => "Soar::Production",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "Exporter" => 0,
    "Parse::RecDescent" => 0,
    "Soar::Production::Parser::PRDGrammar" => 0,
    "base" => 0,
    "parent" => 0,
    "perl" => "5.010",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Soar/Production.pm
#
# This file is part of Soar-Production
#
# This software is copyright (c) 2012 by Nathan Glenn.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
#modified from "Effective Perl Programming" by Joseph N. Hall, et al.
use strict;
use warnings;

# ABSTRACT: REPRESENT SOAR PRODUCTIONS
package Soar::Production;
use Carp;
use Soar::Production::Parser;
use Data::Dumper;
# use Soar::Production::Printer;
use base qw(Exporter);
our @EXPORT_OK = qw(prods_from prods_from_file);

our $VERSION = '0.02'; # VERSION

my $parser = Soar::Production::Parser->new;
# my $printer = Soar::Production::Printer->new;

#if run as a script, prints the name of every production in an input file.
unless(caller){
	my $prods = prods_from(file => $ARGV[0]);
	for my $prod (@$prods){
		print $prod->name . "\n";
	}
}

sub _run {
  my ($prod) = @_;
}

sub new {
  my ($class, $text) = @_;
  my $prod = bless $parser->parse_text($text), $class;
  return $prod;
}

sub as_text {
	my ($class) = @_;
	# return $printer->tree_to_text($class)
}

sub prods_from_file{
	return prods_from( file => shift() );
}

sub prods_from {
	my ($path) = @_;
	my %args = (
		text	=> undef,
		file	=> undef,
		@_
	);
	defined $args{text} or defined $args{file}
		or croak 'Must specify parameter \'file\' or \'text\' to extract productions.';
		
	my $parses = $parser->productions(@_, parse => 1);
	my @prods = map { bless $_ } @$parses;
	
	return \@prods;
}

sub name {
	my ($prod, $name) = @_;
	# print Dumper $prod;
	$prod->{name} = $name
		if($name);
	return $prod->{name};
}

1;

__END__

=pod

=head1 NAME

Soar::Production - REPRESENT SOAR PRODUCTIONS

=head1 VERSION

version 0.02

=head1 SYNOPSIS

=head1 DESCRIPTION

=head1 NAME

Soar::Production- REPRESENTS A SOAR PRODUCTION

=head1 METHODS

=head2 C<new>

Argument: text of a Soar production.
Creates a new production object using the input text.

=head2 C<prods_from>

This method extracts productions from a given text. It returns a reference to an array containing production objects. Note that all comments are removed as a preprocessing step to detecting and extracting productions. It takes a set of named arguments:
'file'- the name of a file to read.
'text'- the text to read.
You must choose to export this function via the C<use> function:

	use Soar::Production qw(prods_from);

=head2 C<prods_from_file>

A shortcut for C<prods_from(file => $arg)>.

=head2 C<name>

Optional argument: name to assign production.
Sets the name of the current production if an argument is given. Returns the name of the production.

=head1 TODO

=head2 C<state_name>

Set/get name of matched state

=head2 C<superstate_name>

Set/get name of matched state's superstate

=head2 C<type>

Does this production match a state or an impasse?

=head2 C<validate>

Check this production against a datamap.

=head2 check semantic correctness

Soar::Production::Parser does not check semantic correctness. The following are good things to check:

=over3

=item everything matched in RHS must be in LHS

=item no empty RHS

=item Only allowable non-operator preference is REJECT

=item Check for existence of RHS function

=item <s> not connected

=item disconnect from goal or impasses (no 'state' or 'impasse' keyword)

=back

=head1 AUTHOR

Nathan Glenn <garfieldnate@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Nathan Glenn.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Soar/Production/Parser.pm
#
# This file is part of Soar-Production
#
# This software is copyright (c) 2012 by Nathan Glenn.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Soar::Production::Parser;
# ABSTRACT: Parse Soar productions

use strict;
use warnings;
#needed for advanced regex expressions
use 5.010;

our $VERSION = '0.02'; # VERSION

use Soar::Production::Parser::PRDGrammar;
use Parse::RecDescent;
use Carp;
use Data::Dumper;

#a regular expression to split text into productions
my $splitter = qr/
	(sp\s+				#start with 'sp'	
	\{					#opening brace
	  (					#save to $2
		 (?: 			#either
			\{ (?-1) \}	#more braces and recurse
			|			#or
			(?:			#group
				[^|{}]++	#not bar or braces
				|   		#or
				\| 			#a bar
				(?:			#group
					[^\\|]++	#no slashes or bars, no backtracking
					|			#or
					\\.			#slash anything
				)*+ 
				\|
			)++				#one or more, no backtracking
		 )*				#0 or more times
	  )					#end $2
	\}					#ending brace
	)					#end $1
/x;

__PACKAGE__->new->_run(shift) unless caller;

sub _run {
  my ($soarParser, $filePath) = @_;
  print Dumper($soarParser->productions(file => $filePath, parse => 1) );
  return;
}

sub new {
  my ($class) = @_;
  my $soarParser = bless {}, $class;
  $soarParser->_init;
  return $soarParser;
}

sub _init {
  my ($soarParser) = @_;
    $soarParser->{parser} = Parse::RecDescent->new($Soar::Production::Parser::PRDGrammar::GRAMMAR);
	
	#if you wish to debug the grammar, try turning on traces by uncommenting the following lines:
	# $::RD_TRACE = 1;
	# $::RD_HINT = 1;
	return;
}


sub productions {## no critic RequireArgUnpacking
    my ($soarParser) = shift;
	my %args = (
		parse	=> 0,
		text	=> undef,
		file	=> undef,
		@_
	);
	defined $args{text} or defined $args{file}
		or croak 'Must specify parameter \'file\' or \'text\' to extract productions.';
	
	if($args{text}){
		return $soarParser->_productions_from_text($args{text}, $args{parse});
	}
	if($args{file}){
		# print "$args{file}\n";
		return $soarParser->_productions_from_file($args{file}, $args{parse});
	}
}

sub _productions_from_text {
    my ( $soarParser, $text, $parse) = @_;
	
	#remove comments
	# print 'Getting prods from ' . $text . "\n";
	$text = $soarParser->no_comment($text);
	# print "no_comments = $text";
	my $productions = _split_text(\$text);
	return $productions
		unless($parse);
	
	return $soarParser->get_parses($productions);
}

sub _productions_from_file {
    my ( $soarParser, $file, $parse) = @_;
    my $text = _readFile($file);
	
	my $productions = _split_text($text);
	return $productions
		unless($parse);
	
	return $soarParser->get_parses($productions);
}

#split text reference into production
sub _split_text {
	my ($text) = @_;
	#split the text into productions by looking for 'sp { ... }'
	my @productions;
	while($$text =~ /$splitter/g){
		# print "found production: $1";
		push @productions, $1;
	}
	return \@productions;
}

sub parse_text {
    my ( $soarParser, $text ) = @_;
    croak 'no text to parse!'
      unless defined $text;
    # $soarParser->{input} = \$text;
    return $soarParser->{parser}->parse($text);
}

#expects array reference holding production text
sub get_parses {
	my ($soarParser, $productions) = @_;
	my @parses;
	for(@$productions){
		# print STDERR $_;
		push @parses, $soarParser->{parser}->parse($_);
	}
	return \@parses;
}

#argument should be an opened file handle
#returns string pointer to text
sub _readFile {
    my ($file) = @_;
    open my $fh, '<', $file
		or croak "Couldn't open $file";
		
    my $text = q();
    $text .= $_ while (<$fh>);
	
	close $fh;
	$text = no_comment(undef,$text);
    return \$text;
}

#remove Soar comments from a line.
sub no_comment {
	#s/#(?=[^|]*$).*$//; #nope
	#s/#.*//g; #nope
	my ($parser, $text) = @_;
	$text =~ s/
			(			#save in $1
				\|			#literal bar
				(?:			#group
					\\[|]		#or an escaped bar
					|
					[^|]		#not a literal bar
				)*		#zero or more of previous group
				\|			#literal bar
			)			#end save
			|			#or
			(?:;\s*)?			# optional semicolon
			\#			# pound character
			.*			#followed by anything
		/			#replace with
			$1||''	# $1 or nothing (the quote if there was one; 
					# no quote will simply remove matching comment)
		/xeg;
	# print "returning $text";
	return $text;
}
1;

__END__

=pod

=head1 NAME

Soar::Production::Parser - Parse Soar productions

=head1 VERSION

version 0.02

=head1 SYNOPSIS

  use Soar::Production::Parser;
  use Data::Dumper;
  my $parser = Soar::Production::Parser->new;
  my @trees=$parser->parse_file("foo.soar");
  print Dumper(\@trees);

=head1 DESCRIPTION

Soar is a cognitive modeling architecture for programming and experimenting with intelligent agents. Soar is programmed using productions that look like this:

	sp{name
		(state <s>)
		-->
		(<s> ^foo bar)
	}

The preceding production matches any state and adds an element named "foo" with the value "bar" to it. Productions can get much more complicated than that.
This module can be used to parse these productions. Underlyingly, a Parse::RecDescent grammar is used to convert a production into a parse tree. 
There are also methods for extracting all of the productions from a file string, and to remove comments (not that I think you'll ever want to do that!). 

=head1 NAME

Soar::Production::Parser - Perl extension for parsing angst grammar files

=head1 METHODS

=head2 C<new>

Creates a new parser.

=head2 C<parse_text>

Argument: the text of a single Soar production.
Returns: a parse tree for the given production.

=head2 C<get_parses>

Argument: Reference to array containing text for individual productions.
Return: Reference to an array containing parse trees for each of the productions in the input array reference.

=head2 C<no_comment>

Argument: Text which contains Soar productions or commands
Return: Same text, but with all comments removed. Comments are indicated with a # (pound), optionally preceded by a ; (semicolon) and whitespace.

=head2 C<productions>

This method extracts productions from a given text. It returns a reference to an array containing either the text of each of the productions, or a parse tree for each of them. Note that all comments are removed as a preprocessing step to detecting and extracting productions. It takes a set of named arguments:
'file'- the name of a file to read.
'text'- the text to split.
'parse'- set to true if the return value should be an array of parse trees for the extracted productions; otherwise an array containing the production text will be returned.
For example, if you would like to extract all of the productions from a file and print their parse trees, you could do this:

    use Soar::Production::Parser;
	use Data::Dumper;
	
	my $file = shift;
	my $parser = Soar::Production::Parser->new();
	my $parses = $parser->productions(
		file => $file,
		parse => 1
	);
	
	for my $prod(@$productions){
		print Dumper($prod);
	}

=head1 SEE ALSO

The documentation for Soar is located at L<https://code.google.com/p/soar/>.
You may also be interested in what a production system is, since this module parses Soar productions: L<http://en.wikipedia.org/wiki/Production_system>.

=head1 AUTHOR

Nathan Glenn <garfieldnate@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Nathan Glenn.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Soar/Production/Printer.pm
#
# This file is part of Soar-Production
#
# This software is copyright (c) 2012 by Nathan Glenn.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
use strict;
use warnings;

# ABSTRACT: Print Soar productions
package Soar::Production::Printer;

use parent 'Exporter';
our @EXPORT = qw(tree_to_text);

use Soar::Production::Parser;
use Data::Dumper;
use Carp;
# use Test::More tests => 1;

our $VERSION = '0.02'; # VERSION

#default behavior is to read the input Soar file and output another one; worthless except for testing
_run(shift) unless caller;

#pass in a Soar grammar file name. Parse the file, then reconstruct it and print to STDOUT.
sub _run {
	my ($file) = @_;

    my $parser = Soar::Production::Parser->new();
    my $trees   = $parser->productions(file => $file, parse => 1);
    croak "parse failure\n" if ( $#$trees == -1 );

    my $text = tree_to_text($$trees[0]);
	my $tree = $parser->parse_text($text)
		or croak 'illegal production printed';
		
	print $text;
	return;
}

sub tree_to_text{
	my ($tree) = @_;
	
    #traverse tree and construct the Soar production text
    my $text = 'sp {';

    $text .= _name( $tree->{name} );
    $text .= _doc( $tree->{doc} );
    $text .= _flags( $tree->{flags} );
	
    $text .= _LHS( $tree->{LHS} );
	$text .= "\n-->\n\t";
    $text .= _RHS( $tree->{RHS} );
	$text .= "\n}";
	
	return $text;
}

sub _name {
	my $name = shift;
	return $name . "\n\t";;
}

sub _doc {
	my $doc = shift;
	if(defined $doc){
		return '"' . $doc . '"' . "\n\t";
	}
	return '';
}

sub _flags {
	my $flags = shift;
	my $text = '';
	for my $flag (@$flags){
		$text .= ':' . $flag . "\n\t";
	}
	return $text;
}

sub _LHS {
	my $LHS = shift;
	return join "\n\t", 
		map { _condition($_) } @{ $LHS->{conditions} };
}

sub _condition {
	my $condition = shift;
	my $text = '';
	
	$text .= '-'
		if($condition->{negative} eq 'yes');
		
	$text .= _positive_condition( $condition->{condition} );
	
	return $text;
}

sub _positive_condition {
	my $condition = shift;
	my $text = '';
	
	return _conjunction( $condition->{conjunction} )
		if($condition->{conjunction});
		
	return _condsForOneId($condition);
}

sub _condsForOneId {
	my $condsForOneId = shift;
	my $text = '(';
	my ($type, $idTest, $attrValueTests) = 
		(
			$condsForOneId->{condType}, 
			$condsForOneId->{idTest}, 
			$condsForOneId->{attrValueTests}
		);
		
	$text .= $type
		if(defined $type);
	
	$text .= ' ' . _test($idTest)
		if(defined $idTest);
		
	if($#$attrValueTests != -1){
		$text .= ' ';
		$text .= join ' ', map { _attrValueTests($_) } @$attrValueTests;
	}
	
	$text .= ')';
	return $text;
}

sub _test {
	my $test = shift;
	
	if(exists $test->{conjunctiveTest}){
		return _conjunctiveTest( 
			$test->{conjunctiveTest} );
	}
	
	return _simpleTest( $test->{simpleTest} );
}

sub _conjunctiveTest {
	my $conjTest = shift;
	my $text = '{';
	$text .= join ' ',
		map { _simpleTest($_) } @$conjTest;
	$text .= '}';
}

sub _simpleTest {
	my $test = shift;
	return _disjunctionTest($test->{disjunctionTest})
		if( exists $test->{disjunctionTest} );
	return _relationalTest($test->{relationTest} )
		if( exists $test->{relationTest} );
	return _singleTest($test);
}

sub _disjunctionTest {
	my $test = shift;
	my $text = '<< ';
	$text .= join ' ', map { _constant($_) } @$test;
	$text .= ' >>';
	return $text;
}

sub _relationalTest {
	my $test = shift;
	
	my $text = _relation( $test->{relation} );
	$text .= ' ';
	$text .= _singleTest( $test->{test} );
	
	return $text;
}

sub _relation {
	my $relation = shift;
	return $relation;
}

sub _singleTest {
	my $test = shift;
	return _variable($test->{variable})
		if( exists $test->{variable} );
	return _constant($test);
}

sub _attrValueTests {
	my $attrValuetests = shift;
	my ($negative, $attrs, $values) = 
		(
			$attrValuetests->{negative},
			$attrValuetests->{attrs},
			$attrValuetests->{values}
		);
	my $text = '';
	$text .= '-'
		if($negative eq 'yes');
	$text .= _attTest($attrs);
	
	if($#$values != -1){
		$text .= ' ';
		$text .= join ' ', map { _valueTest($_) } @$values;
	}
	return $text;
}

sub _attTest {
	my $attTest = shift;
	my $text = '^';
	$text .= join '.', map { _test($_) } @$attTest;
	return $text;
}

sub _valueTest {
	my $valueTest = shift;
	my $text = '';
	
	if(exists $valueTest->{test}){
		$text = _test( $valueTest->{test} );
	}else{
		#condsForOneId
		$text = _condsForOneId($valueTest->{conds});
	}
	
	$text .= '+'
		if($valueTest->{'+'} eq 'yes');
	
	return $text
}

sub _conjunction {
	my $conjunction = shift;
	my $text = '{';
	$text .= join "\n\t", map { _condition($_) } @$conjunction;
	$text .= '}';
	return $text;
}

sub _RHS {
	my $RHS = shift;
	my $text = '';
	for my $action (@$RHS){
		$text .= _action($action);
		$text .= "\n\t";
	}
	return $text;
}

sub _action {
	my $action = shift;
	if(exists $action->{funcCall}){
		return _funcCall($action->{funcCall});
	}
	
	my $text = '(';
	$text .= _variable($action->{variable});
	$text .= ' ';
	$text .= join ' ',
		map {_attrValueMake($_)} @{ $action->{attrValueMake} };
	$text .= ')';
	return $text;
}

sub _attrValueMake {
	my $attrValueMake = shift;
	my ($attr, $valueMake) = 
		($attrValueMake->{attr}, $attrValueMake->{valueMake});
	
	my $text = _attr($$attr[0]);
	if($#$attr != 0){
		$text .= '.';
		$text .= join '.', 
			map { _variableOrSymConstant($_) } @$attr[1..$#$attr];
	}
	
	$text .= ' ';
	$text .= join ' ', map{_valueMake($_)} @$valueMake;
	
	return $text;
}

sub _attr {
	my $attr = shift;
	return '^' . _variableOrSymConstant($attr);
}

sub _variableOrSymConstant {
	my $vOs = shift;
	return _variable($vOs->{variable})
		if(exists $vOs->{variable});
	return _symConstant($vOs);
	
}

sub _valueMake {
	my $valueMake = shift;
	my ($rhsValue, $preferences) = 
		($valueMake->{rhsValue}, $valueMake->{preferences});
	my $text = _rhsValue($rhsValue);
	#there will always be at least one preference; '+' is default
	$text .= ' ';
	$text .= join ',', map { _preference($_) } @$preferences;
	return $text;
}

sub _preference {
	my $preference = shift;
	my $text = $preference->{value};
	if($preference->{type} eq 'binary'){
		$text .= ' ' . _rhsValue( $preference->{compareTo} );
	}
	return $text;
}

#variable | constant | "(crlf)" | funcCall
sub _rhsValue {
	my $rhsValue = shift;
	
	return '(crlf)'
		if($rhsValue eq '(crlf)');
		
	if(exists $rhsValue->{variable}){
		return _variable($rhsValue->{variable});
	}
	if(exists $rhsValue->{constant}){
		return _constant($rhsValue);
	}
	if(exists $rhsValue->{function}){
		return _funcCall($rhsValue);
	}
	return $rhsValue;
}

#(write |Hello World| |hello again|)
sub _funcCall {
	my $funcCall = shift;
	
	my ($name, $args) = 
		(_funcName($funcCall->{function}), $funcCall->{args});
	my $text = '(' . $name;
	if($#$args != -1){
		$text .= ' ';
		$text .= join ' ', map {_rhsValue($_)} @$args;
	}
	return $text . ')';
}

# arithmetic operator (+ - * /) or a symConstant, being the name of some function
sub _funcName {
	my $funcName = shift;
	
	if(ref $funcName eq 'HASH'){
		return _symConstant($funcName);
	}
	return $funcName;
}

sub _variable {
	my $variable = shift;
	return '<' . $variable . '>'
}

sub _constant {
	my $constant = shift;
	my ($type, $value) = ($constant->{type}, $constant->{constant});
	
	return _symConstant($value) if($type eq 'sym');
	return _int($value) if($type eq 'int');
	return _float($value);#only other type is 'float'
}

sub _float {
	my $float = shift;
	return $float;
}

sub _int {
	my $int = shift;
	return $int;
}

#either string or quoted
sub _symConstant {
	my $symConstant = shift;
	my ($type, $value) = ($symConstant->{type}, $symConstant->{value});
	return _string($value) if($type eq 'string');
	return _quoted($value);
}

sub _string {
	return shift;
}

sub _quoted {
	my $text = shift;
	
	#escape vertical bars
	$text =~ s/\|/\\|/g;
	return '|' . $text . '|';
}

1;

__END__

=pod

=head1 NAME

Soar::Production::Printer - Print Soar productions

=head1 VERSION

version 0.02

=head1 SYNOPSIS

  use Soar::Production::Parser;
  use Soar::Production::Printer;
  use Data::Dumper;
  
  #read in a series of productions from a file
  my $parser = Soar::Production::Parser->new;
  my @trees=$parser->parse_file("foo.soar");
  
  #print each of the productions to standard out
  for my $prod(@trees){
	print tree_to_text($prod);
  }

=head1 DESCRIPTION

This module can be used to print production parse trees produced by Soar::Production::parser. Use the function C<tree_to_text> to accomplish this.

Printing is accomplished by traversing the input structure exactly as it is specified by the grammar used by Soar::Production::Parser. 

=head1 NAME

Soar::Production::pRINT - Perl extension for printing Soar productions

=head1 METHODS

=head2 C<tree_to_text>

Argument: parse tree structured as those returned by Soar::Production::Parser.
Returns a text representation of the production which can be sourced by Soar.

=head2 TODO

Pretty printing is not yet possible, which is too bad because it means the output can be pretty disgusting looking.

=head1 AUTHOR

Nathan Glenn <garfieldnate@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Nathan Glenn.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "REPRESENT SOAR PRODUCTIONS\r",
  "AUTHOR" => "Nathan Glenn <garfieldnate\@gmail.com>",
  "BUILD_REQUIRES" => {
    "Data::Diver" => 0,
    "File::Find" => 0,
    "File::Spec" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Module::Build" => "0.3601",
    "Test::Base" => 0,
    "Test::Base::Filter" => 0,
    "Test::Deep" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.88",
    "Test::Warn" => 0,
    "utf8" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Soar-Production",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Soar::Production",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "Exporter" => 0,
    "Parse::RecDescent" => 0,
    "Soar::Production::Parser::PRDGrammar" => 0,
    "base" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.02",
  "test" => {
    "TESTS" => "t/*.t t/parser/*.t t/printer/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



